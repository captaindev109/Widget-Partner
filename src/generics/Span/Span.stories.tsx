import React from 'react';
import { Meta, Story } from '@storybook/react';
import styled from 'styled-components';
import { decorators } from '@/../.storybook/preview';
import Wrapper from '@/generics/Wrapper';
import Span from '.';
import { SpanProps } from '@/types/types';

// A wrapper to position the component
const StoryWrapper = styled(Wrapper)`
  display: flex;
  flex-direction: column;
  gap: 10px;
`;

// Name the folder to either generics or components
const meta: Meta = {
  title: 'Generics/Span',
  component: Span,
  decorators,
};

export default meta;

// Showcasing the component
const Template: Story<SpanProps> = (args) => (
  <StoryWrapper>
    <Span {...args} />
  </StoryWrapper>
);

export const DefaultSpan = Template.bind({});

DefaultSpan.args = {
  children: 'span',
};
