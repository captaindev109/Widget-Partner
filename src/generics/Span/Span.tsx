import { FC } from 'react';
import styled from 'styled-components';
import { SpanProps } from '@/types/types';

const StyledSpan = styled.span<SpanProps>`
  font-size: ${(props) => props.theme.size.fontSizes.f2};
`;

const Span: FC<SpanProps> = ({ children, ...newProps }) => {
  return <StyledSpan {...newProps}>{children}</StyledSpan>;
};

export default Span;
