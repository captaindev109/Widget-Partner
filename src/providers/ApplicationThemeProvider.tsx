import { ThemeProvider, DefaultTheme } from 'styled-components';
import GlobalStyles from '@/styles/Global';
import React from 'react';
import { ChildrenReturnType } from '@/types/types';

interface ApplicationThemeProviderProps {
  children?: ChildrenReturnType;
  theme: DefaultTheme;
}

const ApplicationThemeProvider = ({ children, theme }: ApplicationThemeProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};

export default ApplicationThemeProvider;
