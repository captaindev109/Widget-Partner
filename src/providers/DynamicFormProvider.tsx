import React from 'react';
import { ChildrenReturnType } from '@/types/types';
import { FormProvider, useForm } from 'react-hook-form';

interface DynamicFormProviderProps {
  children?: ChildrenReturnType;
}

const useDynamicRootForm = () => {
  const methods = useForm({ mode: 'onChange' });

  return { methods };
};

export const DynamicFormProvider = ({ children }: DynamicFormProviderProps) => {
  const { methods } = useDynamicRootForm();

  return <FormProvider {...methods}>{children}</FormProvider>;
};
