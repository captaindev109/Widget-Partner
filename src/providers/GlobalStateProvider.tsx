import { mainMachine } from '@/machines/courseMachine';
import headerMachine from '@/machines/headerMachine';
import footerMachine from '@/machines/footerMachine';
import React, { createContext } from 'react';
import { useInterpret } from '@xstate/react';
import { ActorRefFrom } from 'xstate';
import { storeMachine } from '@/machines/storeMachine';

interface GlobalStateContextProps {
  mainService: ActorRefFrom<typeof mainMachine>;
  headerService: ActorRefFrom<typeof headerMachine>;
  footerService: ActorRefFrom<typeof footerMachine>;
  storeService: ActorRefFrom<typeof storeMachine>;
}

export const GlobalStateContext = createContext({} as GlobalStateContextProps);

interface GlobalStateProviderProps {
  children?: React.ReactNode;
}

const useDevTool = process.env.NODE_ENV === 'development';

export const GlobalStateProvider = ({ children }: GlobalStateProviderProps) => {
  const services = {
    mainService: useInterpret(mainMachine, {
      devTools: useDevTool,
    }),
    headerService: useInterpret(headerMachine, {
      devTools: useDevTool,
    }),
    footerService: useInterpret(footerMachine, {
      devTools: useDevTool,
    }),
    storeService: useInterpret(storeMachine, {
      devTools: useDevTool,
    }),
  };

  return <GlobalStateContext.Provider value={services}>{children}</GlobalStateContext.Provider>;
};
