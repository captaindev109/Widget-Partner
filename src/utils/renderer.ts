import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
