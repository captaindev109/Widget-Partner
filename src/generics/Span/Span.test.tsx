/**
 *
 *  ********** Unit & Snapshot Testing *********
 *
 * Unit test is specifiic for each functionliaty of the component
 * Snapshot test is to track Over All changes in the code since the first snap
 **/

import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Span from './Span';
import defaultTheme from '@/theme/theme';
import ApplicationThemeProvider from '@/providers/ApplicationThemeProvider';
import { GlobalStateProvider } from '@/providers/GlobalStateProvider';

const MockComponet: React.FC = () => {
  return (
    <GlobalStateProvider>
      <ApplicationThemeProvider theme={defaultTheme}>
        <Span />
      </ApplicationThemeProvider>
    </GlobalStateProvider>
  );
};

// describe function is a better way to organize our test cases and describe which component we are testing
describe('<Span /> Test Cases', () => {
  // this test is a must for each component, any rendering issues with the component can be found
  test('it renders without crashing, all is well', () => {
    // render basically prints and exposes the component in nodejs environemt, it returns an object...
    render(<MockComponet />);
  });

  it('Snapshot test should pass', () => {
    const snap = renderer.create(<MockComponet />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
