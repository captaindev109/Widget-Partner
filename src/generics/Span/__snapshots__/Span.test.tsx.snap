// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Span /> Test Cases Snapshot test should pass 1`] = `
<span
  className="Span__StyledSpan-sc-1tn16tz-0 ldMWNC"
/>
`;
